feat(missions): Allow missions to complete at alternative locations with `complete at`

UnorderedSigh opened the pull request:
**Feature:** Adds `complete anywhere` and `complete at` properties to missions, which lets them complete at places other than their destinations.

The `complete anywhere` feature was requested by `wanderers start.txt`:
```ruby
# TODO: When "complete anywhere" functionality is available,
# condense the following three missions into a single mission
```

The `complete at` feature is an obvious extension of that, and far more powerful.

## Feature Details

```ruby
mission "Completes at any planet"
    complete anywhere

mission "Completes at any Syndicate non-mining planet near Crossroads"
    complete at
        government Syndicate
        not attributes mining
        near Crossroads 1 3

mission "Completes at any inhabited Remnant planet"
    complete at
        government Remnant
        not attributes uninhabited

mission "Completes at any planet that is 2-7 jumps from the source"
    complete at
        distance 2 7

mission "If `complete at` doesn't match the destination, the mission will not complete there"
    complete at
        government Pirate
    # Until the pirates conquer Earth, this mission won't complete at Earth:
    destination Earth

mission "A `complete at` filter cannot be empty"
    # This is invalid and will be ignored
    complete at
```

Capabilities and limitations:
- It would be wise to hide the destination using this PR: #8778
- An empty `complete at` receives a warning and is ignored.
- A mission will not validate or be offered if the `complete at` filter is invalid.
- If the destination does not match the `complete at` filter, then the mission cannot be completed at the destination.
- All missions must have a destination, even ones with alternative completion locations.
- A mission that can complete the day it starts will complete after the game is loaded. Keep this in mind with `complete anywhere` missions, or missions with destination=source

Other related PR:

- #8770

## Usage Examples
* Escort an NPC to any non-pirate world.
* Kill a pirate ship and report back to any Free Worlds world.
* Carry a passenger to any planet in Remnant space.
* In `wanderers start.txt`, missions `Wanderers Ap'arak 3: Pug 1`, `Wanderers Ap'arak 3: Pug 2`, and `Wanderers Pug Hostile` request this. (See comment at top of PR description.)

## Testing Done

There are three missions to test.

1. Put this in data/[extra-anywhere-test.txt](https://github.com/user-attachments/files/16822186/extra-anywhere-test.txt)
2. Depart.
7. Land at any planet with a spaceport.
8. Visit the spaceport to get three missions: a cargo mission (one ton), a passenger mission (1 bunk) and escorting a merchant Archon.
1. Wait one day.
1. Land on any non-Syndicate planet carrying your passenger to finish one mission.
1. Land on any non-Syndicate planet carrying your cargo to finish a second mission.
1. Land on any Syndicate planet with the Archon you're escorting. This will complete the third mission.

## Performance Impact
N/A

warp-core commented:
I don't think `complete anywhere` and `complete at` (with a location filter) should be handled differently.
A single node can be used for both cases, since any location will match an empty location filter.

Have you checked that a mission using this will still pass `Mission::IsValid()`? It returns false if the destination pointer is null, which will be the case for a mission using this.

UnorderedSigh commented:
> I don't think complete anywhere and complete at (with a location filter) should be handled differently.
> A single node can be used for both cases, since any location will match an empty location filter.

Incorrect. There is no way to tell the difference between no completion filter (ie. only at destination) or complete anywhere (empty completion filter). This way, completeAnywhere=true means the mission completes anywhere, and an empty location filter would indicate no completion information was ever supplied (so, only at destination).

UnorderedSigh commented:
> Have you checked that a mission using this will still pass Mission::IsValid()? It returns false if the destination pointer is null, which will be the case for a mission using this.

Like all missions, a `complete anywhere` mission is still required to have a destination.

UnorderedSigh commented:
I implemented `complete at` and fixed a bug. You can now give completion filters in `complete at` . You can complete anywhere with `complete anywhere` or an empty `complete at` filter.

UnorderedSigh commented:
Last tested and known to work Mon, 29 May 2023 17:29:10 UTC

UnorderedSigh commented:
> No completion filter would be not having a complete at node in the first place no?

I'm referring to how it is stored in the code. You need three possibilities:

1. No completion information was specified at all.
2. A non-empty completion filter was specified.
3. The mission can complete at all locations.

You need a boolean flag, in addition to the filter, because there is no way to tell the difference between option 1 and 3.

UnorderedSigh commented:
Also, typing `complete at` with nothing inside of it doesn't make sense when you read it.

```ruby
mission
    # This makes no sense to a non-coder reading the file:
    complete at
```

If you know how the filter implementation works, then it makes sense. Most content creators don't know how the code works.

UnorderedSigh commented:
I updated the code so an empty `complete at` is an error. You'll get a trace, and the filter will be ignored.

Zitchas commented:
> > No completion filter would be not having a complete at node in the first place no?
> 
> I'm referring to how it is stored in the code. You need three possibilities:
> 
>     1. No completion information was specified at all.
> 
>     2. A non-empty completion filter was specified.
> 
>     3. The mission can complete at all locations.
> 
> 
> You need a boolean flag, in addition to the filter, because there is no way to tell the difference between option 1 and 3.

At the risk of being too simple, what about the possibility of having a special word that can be used in the regular filters that causes the game to interpret it as "all"?

```
mission "Remnant: Logistics 1"
	name "Rush delivery to <planet>"
	description `Deliver <cargo> to <destination> by <date>.`
	source "Caelian"
	destination everywhere
	...
```
Where "everywhere" is a special word (so yes, this would mean we can never name a planet "everywhere") that the game would interpret as matching the first planet the player lands on once they meet the mission completion requirements.

That being said, there's also the need to differentiate between "the selected destination is set to a random planet which could be anywhere in the galaxy" and "the mission will complete on the first planet the player lands on that meets the criteria.

---

Please note, I'm not disagreeing with any of the current proposals, I suppose this is more just looking for elaboration as to why we're leaping to complexity instead of taking the simplest possible approach. (which, admittedly, in the engine mechanics might not actually be the simplest.)

warp-core commented:
> At the risk of being too simple, what about the possibility of having a special word that can be used in the regular filters that causes the game to interpret it as "all"?

This is what an empty filter does: any system or planet will match.
If you simply use the existing destination filter with "match anything," it will just pick a planet (any planet) which is not what is desired here.
The objective is to not pick a planet that matches the filter when the mission is instantiated but instead maintain a filter that planet will be matched against when the mission is ready to be completed. This either requires maintaining a separate filter, or telling the mission to use the existing destination filter object in this different way.

> destination everywhere

This would simply choose a planet called "everywhere" as the destination. Or never instantiate the mission if no such planet exists.

UnorderedSigh commented:
To explain this another way:

1. `destination` = A single location, fixed as of the moment the mission is spawned. This is displayed on the map, and cannot change during the mission. The mission will only be spawned if the destination is valid.
2. `complete at` = A filter for where the mission can *actually complete*. This is evaluated at every landing, so if the game data changes, the completion changes. The `complete at` doesn't even have to match any planets when the mission is spawned.

Here's an example of a `complete at` that matches nothing when the mission is spawned, but can match something later:

```ruby
mission "A rebellion is stirring in the south"
    destination Earth # arbitrary; doesn't matter
    hide destination # another PR allows this
    complete at
        government "Free Worlds"
```

UnorderedSigh commented:
Last tested and known to work Thu, 22 Jun 2023 19:46:17 UTC

warp-core commented on source/Mission.cpp:
```diff
@@ -151,6 +151,20 @@ void Mission::Load(const DataNode &node)
 			deadline = Date(child.Value(1), child.Value(2), child.Value(3));
+		else if(child.Size() == 2 && child.Token(0) == "complete" && child.Token(1) == "anywhere")
+			completeAnywhere = true;
+		else if(child.Size() == 2 && child.Token(0) == "complete" && child.Token(1) == "at")
+		{
+			LocationFilter loaded(child);
+			if(!loaded.IsValid())
+				continue;
+			if(loaded.IsEmpty())
+				child.PrintTrace("Error: The \"complete at\" filter must not be empty. Ignoring this line.");
+			else
+				completionFilter = loaded;
+		}
```
Checking the validity of these filters here probably isn't going to go well.
If they refer to any planets, systems, or outfits that haven't been loaded yet (because, at this point, loading is not complete,) those filters will appear invalid even if they will be valid once everything is fully loaded.

UnorderedSigh replied:
I moved the validity check to Mission::IsValid()

warp-core commented on source/Mission.cpp:
```diff
@@ -151,6 +151,20 @@ void Mission::Load(const DataNode &node)
 			blocked = child.Token(1);
 		else if(child.Token(0) == "deadline" && child.Size() >= 4)
 			deadline = Date(child.Value(1), child.Value(2), child.Value(3));
+		else if(child.Size() == 2 && child.Token(0) == "complete" && child.Token(1) == "anywhere")
+			completeAnywhere = true;
+		else if(child.Size() == 2 && child.Token(0) == "complete" && child.Token(1) == "at")
+		{
+			LocationFilter loaded(child);
+			if(!loaded.IsValid())
+				continue;
+			if(loaded.IsEmpty())
+				child.PrintTrace("Error: The \"complete at\" filter must not be empty. Ignoring this line.");
+			else
+				completionFilter = loaded;
+		}
+		else if(child.Token(0) == "complete")
+			child.PrintTrace("Invalid \"complete\" option.");
```
These should probably go near where the source and destination are loaded (I'd put it immediately after destination.)

Also, I know comparing the size is faster, but that performance difference isn't really important when loading stuff and everything else puts the strings first.
Do:
```c++
else if(child.Token(0) == "complete" && child.Size() == 2)
{
	if(child.Token(1) ==  "anywhere")
		// Whatever
	else if(child.Token(1) == "at")
		// Whatever
	else
		// Print trace
}
```

UnorderedSigh replied:
I moved it after destination, and did something similar to this.

warp-core commented on source/Mission.cpp:
```diff
@@ -346,6 +360,13 @@ void Mission::Save(DataWriter &out, const string &tag) const
 			out.Write("blocked", blocked);
 		if(deadline)
 			out.Write("deadline", deadline.Day(), deadline.Month(), deadline.Year());
+		if(completeAnywhere)
+			out.Write("complete", "anywhere");
+		else if(!completionFilter.IsEmpty() && completionFilter.IsValid())
+		{
+			out.Write("complete", "at");
+			completionFilter.Save(out);
+		}
```
Again, do this right after `destination`.

UnorderedSigh replied:
The deed is done.

UnorderedSigh commented:
The Ubuntu 20.04 GLES test was canceled. It took six hours, so it probably hit the job's wallclock limit.

UnorderedSigh commented:
Last tested and known to work at Sun, 27 Aug 2023 14:45:11 UTC

warp-core commented:
> > Have you checked that a mission using this will still pass Mission::IsValid()? It returns false if the destination pointer is null, which will be the case for a mission using this.
> 
> Like all missions, a `complete anywhere` mission is still required to have a destination.

They probably shouldn't be required to have some specific planet as a destination. It seems rather strange to say "this mission can complete anywhere," and also "the destination of this mission is <planet>."
A mission either being able to complete anywhere or having a valid "complete at" filter should be valid alternatives to having a valid destination planet when checking the validity of a mission.

UnorderedSigh commented:
> > > Have you checked that a mission using this will still pass Mission::IsValid()? It returns false if the destination pointer is null, which will be the case for a mission using this.
> > 
> > 
> > Like all missions, a `complete anywhere` mission is still required to have a destination.
> 
> They probably shouldn't be required to have some specific planet as a destination. It seems rather strange to say "this mission can complete anywhere," and also "the destination of this mission is ." A mission either being able to complete anywhere or having a valid "complete at" filter should be valid alternatives to having a valid destination planet when checking the validity of a mission.

That would require a major refactoring of the mission code. Removing the necessity of a destination should be in a separate PR.

UnorderedSigh commented:
If you want to eliminate the need for a destination, I suggest you do this:

1. Merge this PR so we have a viable starting point.
2. Create a github project to track the PRs that solve the individual problems. From skimming the code, I expect it'll take about ten PRs.

UnorderedSigh commented:
Last tested and known to work Wed, 22 Nov 2023 23:11:51 UTC

UnorderedSigh commented:
Last tested and known to work Tue, 27 Aug 2024 17:45:23 UTC.

tibetiroka commented:
People have been requesting for a long time that we should be able to complete missions on the fly (in a dialog/conversation) instead of failing them, because having intentionally failed missions in the save file can confuse people. Would that feature, combined with standard mission actions and filters, be able to replicate this feature?

UnorderedSigh commented:
> People have been requesting for a long time that we should be able to complete missions on the fly (in a dialog/conversation) instead of failing them, because having intentionally failed missions in the save file can confuse people. Would that feature, combined with standard mission actions and filters, be able to replicate this feature?

That would be a wonderful, glorious feature, that I truly desire. Unfortunately, it won't solve all the problems.

Suppose the mission requires escorting an NPC or carrying cargo. There's no way to determine if that worked other than letting the game run its own full completion checks.

I just realized I didn't test that with my changes. Do they correctly handle NPCs, cargo, and passengers? I don't know... I'll have to check.

UnorderedSigh commented:
I've tested my changes with a passenger mission, a cargo mission, and an NPC escort mission. They work!

petervdmeer commented:
I merged the master branch for this PR, and I limited the feature to `complete at`, since everyone seems to agree on that part of the PR.

My intention is to merge `complete at` through this PR, and that we get a follow-up PR for `complete anywhere`, another PR for the missions without destinations, and then other ten PRs that are mentioned earlier to solve individual problems.
